runtime/vcache/project.go

```diff
@@ -0,0 +1,122 @@
+package vcache
+
+import (
+	"fmt"
+
+	"github.com/brimdata/zed"
+	"github.com/brimdata/zed/vector"
+)
+
+func project(zctx *zed.Context, paths Path, s shadow) vector.Any {
+	switch s := s.(type) {
+	case *variant:
+		return projectVariant(zctx, paths, s)
+	case *record:
+		return projectRecord(zctx, paths, s)
+	case *array:
+		vals := project(zctx, nil, s.vals)
+		typ := zctx.LookupTypeArray(vals.Type())
+		return vector.NewArray(typ, s.offs, vals, s.nulls.flat)
+	case *set:
+		vals := project(zctx, nil, s.vals)
+		typ := zctx.LookupTypeSet(vals.Type())
+		return vector.NewSet(typ, s.offs, vals, s.nulls.flat)
+	case *map_:
+		keys := project(zctx, nil, s.keys)
+		vals := project(zctx, nil, s.vals)
+		typ := zctx.LookupTypeMap(keys.Type(), vals.Type())
+		return vector.NewMap(typ, s.offs, keys, vals, s.nulls.flat)
+	case *union:
+		return projectUnion(zctx, nil, s)
+	case *primitive:
+		if len(paths) > 0 {
+			return vector.NewMissing(zctx, s.length())
+		}
+		return s.vec
+	case *const_:
+		if len(paths) > 0 {
+			return vector.NewMissing(zctx, s.length())
+		}
+		return s.vec
+	case *named:
+		v := project(zctx, paths, s.vals)
+		typ, _ := zctx.LookupTypeNamed(s.name, v.Type())
+		return vector.NewNamed(typ, v)
+	default:
+		panic(fmt.Sprintf("vector cache: shadow type %T not supported", s))
+	}
+}
+
+func projectVariant(zctx *zed.Context, paths Path, s *variant) vector.Any {
+	vecs := make([]vector.Any, 0, len(s.vals))
+	for _, m := range s.vals {
+		vec := project(zctx, paths, m)
+		vecs = append(vecs, vec)
+	}
+	return vector.NewVariant(s.tags, vecs)
+}
+
+func projectRecord(zctx *zed.Context, paths Path, s *record) vector.Any {
+	if len(paths) == 0 {
+		// Build the whole record.  We're either loading all on demand (nil paths)
+		// or loading this record because it's referenced at the end of a projected path.
+		vals := make([]vector.Any, 0, len(s.fields))
+		types := make([]zed.Field, 0, len(s.fields))
+		for _, f := range s.fields {
+			v := project(zctx, nil, f.val)
+			vals = append(vals, v)
+			types = append(types, zed.Field{Name: f.name, Type: v.Type()})
```

Comment:
```suggestion
			vec := project(zctx, nil, f.val)
			vals = append(vals, vec)
			types = append(types, zed.Field{Name: f.name, Type: vec.Type()})
```

Reply:
I'm going with val here for reason described above